src/encoding.rs

```diff
@@ -0,0 +1,197 @@
+/// |-------------|------------------------------------------
+/// |`00 00 FE FF`|UCS-4, big-endian machine (1234 order)
+/// |`FF FE 00 00`|UCS-4, little-endian machine (4321 order)
+/// |`00 00 FF FE`|UCS-4, unusual octet order (2143)
+/// |`FE FF 00 00`|UCS-4, unusual octet order (3412)
+/// |`FE FF ## ##`|UTF-16, big-endian
+/// |`FF FE ## ##`|UTF-16, little-endian
+/// |`EF BB BF`   |UTF-8
+/// |-------------|------------------------------------------
+/// |`00 00 00 3C`|UCS-4 or similar (use declared encoding to find the exact one), in big-endian (1234)
+/// |`3C 00 00 00`|UCS-4 or similar (use declared encoding to find the exact one), in little-endian (4321)
+/// |`00 00 3C 00`|UCS-4 or similar (use declared encoding to find the exact one), in unusual byte orders (2143)
+/// |`00 3C 00 00`|UCS-4 or similar (use declared encoding to find the exact one), in unusual byte orders (3412)
+/// |`00 3C 00 3F`|UTF-16 BE or ISO-10646-UCS-2 BE or similar 16-bit BE (use declared encoding to find the exact one)
+/// |`3C 00 3F 00`|UTF-16 LE or ISO-10646-UCS-2 LE or similar 16-bit LE (use declared encoding to find the exact one)
+/// |`3C 3F 78 6D`|UTF-8, ISO 646, ASCII, some part of ISO 8859, Shift-JIS, EUC, or any other 7-bit, 8-bit, or mixed-width encoding which ensures that the characters of ASCII have their normal positions, width, and values; the actual encoding declaration must be read to detect which of these applies, but since all of these encodings use the same bit patterns for the relevant ASCII characters, the encoding declaration itself may be read reliably
+/// |`4C 6F A7 94`|EBCDIC (in some flavor; the full encoding declaration must be read to tell which code page is in use)
+/// |_Other_      |UTF-8 without an encoding declaration, or else the data stream is mislabeled (lacking a required encoding declaration), corrupt, fragmentary, or enclosed in a wrapper of some kind
+///
+/// Because [`encoding_rs`] crate supported only subset of those encodings, only
+/// supported subset are detected, which is UTF-8, UTF-16 BE and UTF-16 LE.
+///
+/// If encoding is detected, `Some` is returned, otherwise `None` is returned.
+#[cfg(feature = "encoding")]
+pub fn detect_encoding(bytes: &[u8]) -> Option<&'static Encoding> {
+    match bytes {
+        // with BOM
+        _ if bytes.starts_with(&[0xFE, 0xFF]) => Some(UTF_16BE),
+        _ if bytes.starts_with(&[0xFF, 0xFE]) => Some(UTF_16LE),
+        _ if bytes.starts_with(&[0xEF, 0xBB, 0xBF]) => Some(UTF_8),
+
+        // without BOM
+        _ if bytes.starts_with(&[0x00, b'<', 0x00, b'?']) => Some(UTF_16BE), // Some BE encoding, for example, UTF-16 or ISO-10646-UCS-2
+        _ if bytes.starts_with(&[b'<', 0x00, b'?', 0x00]) => Some(UTF_16LE), // Some LE encoding, for example, UTF-16 or ISO-10646-UCS-2
+        _ if bytes.starts_with(&[b'<', b'?', b'x', b'm']) => Some(UTF_8), // Some ASCII compatible
+
+        _ => None,
+    }
+}
+
+// TODO: add some tests for functions
```

Comment:
Don't mind to add tests before merge?

Reply:
I considered it but I figured it will be easier to do a big testing push at the end.  We've only got a few sample documents and entering the data manually would be painful.

Reply:
Of course, it needs to happen eventually - it would just be helpful to have the full picture in mind while doing that work.